src/optimizer/statistics/stats_calculator.cpp

```diff
@@ -37,231 +35,112 @@ void StatsCalculator::Visit(const LogicalGet *op) {
     return;
   }
 
-  auto root_group = context_->GetMemo().GetGroupByID(gexpr_->GetGroupID());
-  auto table_stats = context_->GetStatsStorage()->GetTableStats(op->GetDatabaseOid(), op->GetTableOid());
-  if (table_stats == nullptr) {
-    // no table stats
-    // Fill with defaults to prevent an infinite loop from above
-    for (auto &col : required_cols_) {
-      NOISEPAGE_ASSERT(col->GetExpressionType() == parser::ExpressionType::COLUMN_VALUE, "Expected ColumnValue");
-      auto tv_expr = col.CastManagedPointerTo<parser::ColumnValueExpression>();
-      root_group->AddStats(tv_expr->GetFullName(), CreateDefaultStats(tv_expr));
-    }
-    return;
-  }
-
-  // First, get the required stats of the base table
-  std::unordered_map<std::string, std::unique_ptr<ColumnStats>> required_stats;
-  for (auto &col : required_cols_) {
-    // Make a copy for required stats since we may want to modify later
-    AddBaseTableStats(col, table_stats, &required_stats);
-  }
+  auto *root_group = context_->GetMemo().GetGroupByID(gexpr_->GetGroupID());
 
   // Compute selectivity at the first time
   if (root_group->GetNumRows() == -1) {
-    std::unordered_map<std::string, std::unique_ptr<ColumnStats>> predicate_stats;
-    for (auto &annotated_expr : op->GetPredicates()) {
-      ExprSet expr_set;
-      auto predicate = annotated_expr.GetExpr();
-      parser::ExpressionUtil::GetTupleValueExprs(&expr_set, predicate);
-      for (auto &col : expr_set) {
-        AddBaseTableStats(col, table_stats, &predicate_stats);
-      }
-    }
-
-    // Use predicates to estimate cardinality. If we were unable to find any column stats from the catalog, default to 0
-    if (table_stats->GetColumnCount() == 0) {
-      root_group->SetNumRows(0);
-    } else {
-      auto est = EstimateCardinalityForFilter(table_stats->GetNumRows(), predicate_stats, op->GetPredicates());
-      root_group->SetNumRows(static_cast<int>(est));
-    }
-  }
+    const auto &stats_cache_reference = context_->GetStatsStorage()->GetTableStats(
```

Comment:
I don't think the `&` is safe here. The return value of `GetTableStats` is an `rvalue` (value on the stack) and just taking a `&` here would not store it anywhere. The value can be corrupted. I think we should use `const auto` here and in similar places (e.g., in `stats_storage_test`).

Reply:
Fixed